### PR TITLE
Take msgv::Value by value in bridge::events

### PR DESCRIPTION
**Pull request by luisholanda**
This PR applies the suggestion made by @KillTheMule in #81, i.e. remove the rest of clones by taking ownership of `Value`.

In the end, the code is cleaner than before but I needed to use a macro to mimic the slice pattern (but taking ownership of the elements). I'm open to rewriting if anyone knows a better way to do something like that.

### CODE REVIEW

File: src/bridge/events.rs [hunk 1]
```diff
@@ -159,494 +160,455 @@ fn unpack_color(packed_color: u64) -> Color4f {
...
+fn parse_map(map_value: Value) -> Result<Vec<(Value, Value)>> {
+    map_value.try_into().map_err(EventParseError::InvalidMap)
 }
 
-fn parse_i64(i64_value: &Value) -> Result<i64> {
-    if let Value::Integer(content) = i64_value {
-        Ok(content.as_i64().ok_or_else(|| EventParseError::InvalidI64(i64_value.clone()))?)
-    } else {
-        Err(EventParseError::InvalidI64(i64_value.clone()))
+fn parse_string(string_value: Value) -> Result<String> {
```
Comment (KillTheMule) [inline src/bridge/events.rs]: This could use https://docs.rs/rmpv/0.4.4/rmpv/enum.Value.html#impl-TryFrom%3CValue%3E

**Comment by KillTheMule** [comment]
> I needed to use a macro to mimic the slice pattern

Well, you don't realy _need_ to, you could just write the code at every callsite. It's doing exactly what one would do manually, after all (except it needs to do this weird workarouns with the enum). I personally find it a bit to magical for my taste and I'd just write the code out in the parsing functions.

**Comment by luisholanda** [comment]
> (e) Actually, it's used quite a bit, so maybe I'd still keep it this way.

Yeah, I started writing down in each location but there was so much duplication that I replaced with a macro.

**Comment by luisholanda** [comment]
Humm, this PR is not ready to merge, something in it segfaulting in MacOS.

This is the call stack that I receive:

```
2020-01-31 08:58:17.640 neovide[64129:5427753] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'NSWindow drag regions should only be invalidated on the Main Thread!'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff4367bacd __exceptionPreprocess + 256
	1   libobjc.A.dylib                     0x00007fff6dd7da17 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff43695629 -[NSException raise] + 9
	3   AppKit                              0x00007fff40c3a5ca -[NSWindow(NSWindow_Theme) _postWindowNeedsToResetDragMarginsUnlessPostingDisabled] + 317
	4   AppKit                              0x00007fff40c379f7 -[NSWindow _initContent:styleMask:backing:defer:contentView:] + 1479
	5   AppKit                              0x00007fff40cf6d95 -[NSPanel _initContent:styleMask:backing:defer:contentView:] + 50
	6   AppKit                              0x00007fff40c3742a -[NSWindow initWithContentRect:styleMask:backing:defer:] + 45
	7   AppKit                              0x00007fff40cf6d4a -[NSPanel initWithContentRect:styleMask:backing:defer:] + 64
	8   AppKit                              0x00007fff40c3596f -[NSWindowTemplate nibInstantiate] + 484
	9   AppKit                              0x00007fff40bb2470 -[NSIBObjectData instantiateObject:] + 267
	10  AppKit                              0x00007fff40bb1805 -[NSIBObjectData nibInstantiateWithOwner:options:topLevelObjects:] + 578
	11  AppKit                              0x00007fff40bafa60 loadNib + 401
	12  AppKit                              0x00007fff40baef40 +[NSBundle(NSNibLoading) _loadNibFile:nameTable:options:withZone:ownerBundle:] + 696
	13  AppKit                              0x00007fff40baeb85 -[NSBundle(NSNibLoading) loadNibNamed:owner:topLevelObjects:] + 204
	14  AppKit                              0x00007fff40f601de -[NSAlert init] + 101
	15  neovide                             0x0000000101b5ee0a _ZN60_$LT$$LP$$RP$$u20$as$u20$objc..message..MessageArguments$GT$6invoke17h2a194ef72bba4045E + 58
	16  neovide                             0x0000000101b5bc2e _ZN4objc7message8platform15send_unverified17h0213ffa84d3ff958E + 110
	17  neovide                             0x000000010130e4d8 _ZN76_$LT$$BP$mut$u20$objc..runtime..Object$u20$as$u20$msgbox..macos..NSAlert$GT$4init17h51c2b037a23c2c48E + 280
	18  neovide                             0x000000010130f670 _ZN6msgbox5macos6create17hd87234d2ba0f21acE + 112
	19  neovide                             0x00000001011caeb6 _ZN7neovide14error_handling10show_error17h5a53ab6357940c1cE + 54
	20  neovide                             0x000000010121a32a _ZN114_$LT$core..result..Result$LT$T$C$E$GT$$u20$as$u20$neovide..error_handling..ResultPanicExplanation$LT$T$C$E$GT$$GT$25unwrap_or_explained_panic17h5b567c7cc63b1c56E + 586
	21  neovide                             0x00000001012b99c0 _ZN90_$LT$neovide..bridge..handler..NeovimHandler$u20$as$u20$nvim_rs..rpc..handler..Handler$GT$13handle_notify15__handle_notify28_$u7b$$u7b$closure$u7d$$u7d$17h3a966fdded6f2133E + 400
	22  neovide                             0x00000001012b3824 _ZN80_$LT$std..future..GenFuture$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$4poll17h1161df6533d2a377E + 68
	23  neovide                             0x00000001011e745b _ZN72_$LT$core..pin..Pin$LT$P$GT$$u20$as$u20$core..future..future..Future$GT$4poll17hba47cd5ac14b15f3E + 107
	24  neovide                             0x00000001012b2cdd _ZN3std6future21poll_with_tls_context17hbcd9e6d47aa00b02E + 141
	25  neovide                             0x0000000101303260 _ZN7nvim_rs6neovim15Neovim$LT$W$GT$7io_loop28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$17hce61780c26fb19a4E + 304
	26  neovide                             0x00000001012b5b04 _ZN80_$LT$std..future..GenFuture$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$4poll17he2b89b06d639dfb7E + 68
	27  neovide                             0x000000010124c984 _ZN5tokio4task4core13Core$LT$T$GT$4poll17h16be72033f84cc2bE + 388
	28  neovide                             0x0000000101296732 _ZN5tokio4task7harness20Harness$LT$T$C$S$GT$4poll28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$17h9eb60cfd3b251600E + 50
	29  neovide                             0x00000001012d6509 _ZN4core3ops8function6FnOnce9call_once17h287c5dfd45cff8b7E + 25
	30  neovide                             0x00000001012e31ed _ZN101_$LT$std..panic..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$9call_once17h700625d4a28ed83fE + 29
	31  neovide                             0x00000001012271ca _ZN3std9panicking3try7do_call17h1a91986b5815b65fE + 58
	32  neovide                             0x0000000101b9affb __rust_maybe_catch_panic + 27
	33  neovide                             0x000000010122583b _ZN3std9panicking3try17h50842f51a4ce72a6E + 155
	34  neovide                             0x00000001012e3914 _ZN3std5panic12catch_unwind17hb61ab5e9c4f1cf6bE + 36
	35  neovide                             0x0000000101296158 _ZN5tokio4task7harness20Harness$LT$T$C$S$GT$4poll28_$u7b$$u7b$closure$u7d$$u7d$17hd71e8efc643c4895E + 72
	36  neovide                             0x00000001012ef0da _ZN5tokio4loom3std11causal_cell19CausalCell$LT$T$GT$8with_mut17hc3bde62cb7cd9731E + 106
	37  neovide                             0x0000000101290700 _ZN5tokio4task7harness20Harness$LT$T$C$S$GT$4poll17h05265fb59bf791ffE + 752
	38  neovide                             0x00000001012f13c2 _ZN5tokio4task3raw4poll17hdbaf235925f575bcE + 50
	39  neovide                             0x00000001015ad59a _ZN5tokio4task3raw7RawTask4poll17h825908ccf46731cdE + 90
	40  neovide                             0x000000010158c464 _ZN5tokio4task13Task$LT$S$GT$3run17h85ab34c7cd84ec86E + 52
	41  neovide                             0x000000010157554a _ZN5tokio7runtime11thread_pool6worker15GenerationGuard8run_task17hd6c604f1e8af6ef9E + 154
	42  neovide                             0x0000000101574d84 _ZN5tokio7runtime11thread_pool6worker15GenerationGuard22process_available_work17h32e9507d375830e0E + 516
	43  neovide                             0x0000000101574ab0 _ZN5tokio7runtime11thread_pool6worker15GenerationGuard3run17hbcb7228e147023b0E + 64
	44  neovide                             0x0000000101574679 _ZN5tokio7runtime11thread_pool6worker6Worker3run28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$17heae9e621d329312bE + 169
	45  neovide                             0x00000001015e5a83 _ZN3std6thread5local17LocalKey$LT$T$GT$8try_with17ha15e12e38e3666e1E + 211
	46  neovide                             0x00000001015e4fb5 _ZN3std6thread5local17LocalKey$LT$T$GT$4with17hcea68929af42a96dE + 37
	47  neovide                             0x0000000101574706 _ZN5tokio7runtime11thread_pool6worker6Worker3run28_$u7b$$u7b$closure$u7d$$u7d$17hbd6fa4ee803c6c2fE + 70
	48  neovide                             0x00000001015e20b4 _ZN5tokio7runtime11thread_pool7current3set28_$u7b$$u7b$closure$u7d$$u7d$17h66ce63f32a27bac1E + 212
	49  neovide                             0x00000001015e5dbd _ZN3std6thread5local17LocalKey$LT$T$GT$8try_with17hb1ebbb4c3091b052E + 237
	50  neovide                             0x00000001015e4b78 _ZN3std6thread5local17LocalKey$LT$T$GT$4with17h06e13fdc66f8329dE + 56
	51  neovide                             0x00000001015e1fd8 _ZN5tokio7runtime11thread_pool7current3set17h5807dc7c0febf12bE + 72
	52  neovide                             0x00000001015744f3 _ZN5tokio7runtime11thread_pool6worker6Worker3run17he7a6e11b7b925aa5E + 147
	53  neovide                             0x0000000101584591 _ZN5tokio7runtime11thread_pool7Workers5spawn28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$17h0f649be1e87783d5E + 49
	54  neovide                             0x0000000101588c22 _ZN102_$LT$tokio..runtime..blocking..task..BlockingTask$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$4poll17ha0634f5773435c00E + 178
	55  neovide                             0x0000000101576cb3 _ZN5tokio4task4core13Core$LT$T$GT$4poll17h9a55b19112d55fbbE + 355
	56  neovide                             0x000000010156b592 _ZN5tokio4task7harness20Harness$LT$T$C$S$GT$4poll28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$17hdf831e8fd6fa6f6cE + 50
	57  neovide                             0x00000001015cc8f9 _ZN4core3ops8function6FnOnce9call_once17h78aa5a8ae226c24dE + 25
	58  neovide                             0x0000000101598f2d _ZN101_$LT$std..panic..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$9call_once17hb59ab8121fa2bd5cE + 29
	59  neovide                             0x00000001015b18da _ZN3std9panicking3try7do_call17h19740465a78f384cE + 58
	60  neovide                             0x0000000101b9affb __rust_maybe_catch_panic + 27
	61  neovide                             0x00000001015b156b _ZN3std9panicking3try17h698799f7df265a7eE + 155
	62  neovide                             0x00000001015991e4 _ZN3std5panic12catch_unwind17hc95accf5e65995b4E + 36
	63  neovide                             0x000000010156b638 _ZN5tokio4task7harness20Harness$LT$T$C$S$GT$4poll28_$u7b$$u7b$closure$u7d$$u7d$17h03ab2560f40176bdE + 72
	64  neovide                             0x000000010158274a _ZN5tokio4loom3std11causal_cell19CausalCell$LT$T$GT$8with_mut17h655489a52589aeeeE + 106
	65  neovide                             0x000000010156b1a0 _ZN5tokio4task7harness20Harness$LT$T$C$S$GT$4poll17h39af61cdaaa46a8aE + 752
	66  neovide                             0x00000001015ad852 _ZN5tokio4task3raw4poll17h528cdfdf2c59eeceE + 50
	67  neovide                             0x00000001015ad59a _ZN5tokio4task3raw7RawTask4poll17h825908ccf46731cdE + 90
	68  neovide                             0x000000010158c504 _ZN5tokio4task13Task$LT$S$GT$3run17hc2afff7359fb13fdE + 52
	69  neovide                             0x00000001015e1e9c _ZN5tokio7runtime8blocking4pool8run_task17hc6d41047d4e0d557E + 44
	70  neovide                             0x00000001015e16fd _ZN5tokio7runtime8blocking4pool5Inner3run17hd531e7f9f4ef20fdE + 397
	71  neovide                             0x00000001015e14a7 _ZN5tokio7runtime8blocking4pool7Spawner12spawn_thread28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$17h7b217db13e2866f6E + 55
	72  neovide                             0x00000001015a7d96 _ZN5tokio7runtime7context5enter17h9a74e84b7aa58247E + 86
	73  neovide                             0x00000001015836cb _ZN5tokio7runtime6handle6Handle5enter17h54b778c8b6431031E + 75
	74  neovide                             0x00000001015e1530 _ZN5tokio7runtime8blocking4pool7Spawner12spawn_thread28_$u7b$$u7b$closure$u7d$$u7d$17h01500f91a98a0177E + 48
	75  neovide                             0x00000001015b0af1 _ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17hd47c644555ad50d2E + 49
	76  neovide                             0x00000001015a3e11 _ZN3std6thread7Builder15spawn_unchecked28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$17h51ddf553f255340aE + 49
	77  neovide                             0x0000000101598f91 _ZN101_$LT$std..panic..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$9call_once17hcde1f627bdfe3001E + 49
	78  neovide                             0x00000001015b1973 _ZN3std9panicking3try7do_call17h719d95e3016fb0a2E + 99
	79  neovide                             0x0000000101b9affb __rust_maybe_catch_panic + 27
	80  neovide                             0x00000001015b16fa _ZN3std9panicking3try17he789bb83e5bf9dc5E + 234
	81  neovide                             0x0000000101599171 _ZN3std5panic12catch_unwind17h9f5920d7e8dd9e52E + 49
	82  neovide                             0x00000001015a3c4f _ZN3std6thread7Builder15spawn_unchecked28_$u7b$$u7b$closure$u7d$$u7d$17h84a37b3946ec349dE + 287
	83  neovide                             0x00000001015cc665 _ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17h2584708ba4a1bbd7E + 21
	84  neovide                             0x0000000101b8ae0e _ZN83_$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$9call_once17hc4b783f02b6030e4E + 62
	85  neovide                             0x0000000101b9aa0e _ZN3std3sys4unix6thread6Thread3new12thread_start17h7596958dce7451a0E + 142
	86  libsystem_pthread.dylib             0x00007fff6f7402eb _pthread_body + 126
	87  libsystem_pthread.dylib             0x00007fff6f743249 _pthread_start + 66
	88  libsystem_pthread.dylib             0x00007fff6f73f40d thread_start + 13
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

**Comment by luisholanda** [comment]
I fixed what was causing the call to `error_handling::show_error` but I didn't try to fix the problem with `msgbox::create`, as I think is out of the scope of this PR. I will open an issue for this.

File: src/bridge/events.rs [hunk 2]
```diff
@@ -148,7 +149,7 @@ pub enum RedrawEvent {
 
 fn unpack_color(packed_color: u64) -> Color4f {
     let packed_color = packed_color as u32;
-    let r = ((packed_color & 0xff0000) >> 16) as f32;
+    let r = ((packed_color & 0x00ff_0000) >> 16) as f32;
```
Comment (Kethku) [inline src/bridge/events.rs]: nit: the underscore added here obscures the pattern of accessing the different hex values as color. I'd prefer it not be changed.

File: src/bridge/events.rs [hunk 3]
```diff
@@ -159,494 +160,458 @@ fn unpack_color(packed_color: u64) -> Color4f {
...
+    values.pop().ok_or(EventParseError::InvalidEventFormat)
 }
 
-fn parse_map(map_value: &Value) -> Result<&[(Value, Value)]> {
-    if let Value::Map(content) = map_value {
-        Ok(&content[..])
-    } else {
-        Err(EventParseError::InvalidMap(map_value.clone()))
+macro_rules! extract_values {
+    ([$($i: ident),+], $values: expr $(,)?) => {
```
Comment (Kethku) [inline src/bridge/events.rs]: I feel pretty strongly that we should only use macros when there is no obvious way to implement the functionality without them or if the readability of the code goes significantly up.

In this case there is a clear and obvious readability gain, but I'm not sure that we couldn't have solved this with a simple function. For example, what if instead we wrote a function which took a value and an expected number of arguments and returned a result which is the array of that number of arguments if the count matches, or the appropriate error otherwise.

Such an function could be used like this:

``` Rust
let [_cursor_style_enabled, mode_info] = extract_values(mode_info_set_arguments, 2)?;
```

Then when someone comes along to read the code they fully understand the contract and when jumping to the definition, its familiar code instead of the relatively hard to parse macro syntax. Does that make sense?

File: src/bridge/events.rs [hunk 4]
```diff
@@ -159,494 +160,458 @@ fn unpack_color(packed_color: u64) -> Color4f {
...
+            .map(parse_msg_history_entry)
+            .collect::<Result<_>>()?
+    })
 }
 
-pub fn parse_redraw_event(event_value: &Value) -> Result<Vec<RedrawEvent>> {
-    let event_contents = parse_array(event_value)?;
-    let name_value = event_contents.get(0).ok_or(EventParseError::InvalidEventFormat)?;
+pub fn parse_redraw_event(event_value: Value) -> Result<Vec<RedrawEvent>> {
+    let mut event_contents = parse_array(event_value)?.into_iter();
```
Comment (Kethku) [inline src/bridge/events.rs]: This is a micro optimization which I think obscures the original intention (take the first for the name and the rest for the arguments) for little observable performance gain. Event parsing doesn't show up on the profiling of the app as an appreciable performance concern, so I would rather leave it as it makes the code more readable.

I realize that the previous code took it as a reference, so I'm not sure what the exact right strategy is here, but if at all possible I would like to preserve the initial reading where it is obvious that the name is the first and the arguments are the rest. This could be achieved just with properly named variables... Not sure.

That said, this is relatively nit picky so take it or leave it.

File: src/bridge/events.rs (hunk 2 above)
Comment (Kethku) [reply src/bridge/events.rs]: Alternatively you could add an underscore between every pair of 2 hex values since thats how the colors are read. This isn't a number in the traditional sense

File: src/bridge/events.rs (hunk 3 above)
Comment (luisholanda) [reply src/bridge/events.rs]: Humm, maybe I can do something as you suggested. It wouldn't be exactly as you suggested, as the size of an array should be known at compile-time, but I think we should be able to write a function that does this.

Comment (Kethku) [reply src/bridge/events.rs]: Maybe a dumb question, but why does the size of the array need to be known at compile time? Oh. So the de-structuring works?

Comment (luisholanda) [reply src/bridge/events.rs]: > why does the size of the array need to be known at compile time?

Because the compiler needs to know the size of the array to know how much memory must the reserved for it on the stack.

File: src/bridge/events.rs (hunk 4 above)
Comment (luisholanda) [reply src/bridge/events.rs]: Actually, I didn't do it as an optimization 😅, was due to the fact that we need ownership of the values.

> think obscures the original intention (take the first for the name and the rest for the arguments)

Actually, this code still expresses this intention, as we take the first element of the iterator (`event_contents.next()`) and use the rest of it for the arguments (`events = event_contents`).

> This could be achieved just with properly named variables... Not sure.

Maybe, I will try to find better names.

**Comment by luisholanda** [comment]
I was able to remove the macro from the code. We could improve this code further if `rmpv::Value` implemented `Default` but I think what we have now is much better than the macro version.

File: src/bridge/events.rs (hunk 2 above)
Comment (luisholanda) [reply src/bridge/events.rs]: I decided to keep the old format.

**Comment by Kethku** [comment]
Thanks for the contribution! I really appreciate it :)
